### PR TITLE
[WFLY-14255] allow configuration of reuse-x-forwarded and rewrite-host

### PR DESCRIPTION
**Pull request by baranowb**
Issue: https://issues.redhat.com/browse/WFLY-14255
Replaces: https://github.com/wildfly/wildfly/pull/14859

____

More information about the [wildfly-bot[bot]](https://github.com/wildfly/wildfly-github-bot)

### CODE REVIEW

**Review by fl4via — APPROVED** [review]

**Review by ropalka — APPROVED** [review]

**Comment by baranowb** [comment]
AFAIR, reqs are in, but just to have it go through whole cycle Ive recreated labels from OG PR.

**Comment by baranowb** [comment]
Remove tag after rebase.

**Review by fl4via — APPROVED** [review]

**Review by fl4via — APPROVED** [review]

File: undertow/src/main/java/org/wildfly/extension/undertow/UndertowSubsystemSchema.java [hunk 1]
```diff
@@ -34,10 +39,11 @@ public enum UndertowSubsystemSchema implements PersistentSubsystemSchema<Underto
     VERSION_11_0(11),   // WildFly 20-22    N.B. There were no parser changes between 10.0 and 11.0 !!
     VERSION_12_0(12),   // WildFly 23-26.1, EAP 7.4
     VERSION_13_0(13),   // WildFly 27       N.B. There were no schema changes between 12.0 and 13.0!
-    VERSION_14_0(14),   // WildFly 28-present
+    VERSION_14_0(14),   // WildFly 28
+    VERSION_15_0_PREVIEW(15, 0, Stability.PREVIEW)   // WildFly 32-present
```
Comment (PrarthonaPaul) [inline undertow/src/main/java/org/wildfly/extension/undertow/UndertowSubsystemSchema.java:43]: I think the new schema version should be `Version_14_preview`, `version_14_community` etc. 
If you add another preview schema once this RFE is released, then the new one would be `15:preview`

File: undertow/src/test/java/org/wildfly/extension/undertow/AbstractUndertowSubsystemTestCase.java [hunk 2]
```diff
@@ -34,14 +34,10 @@
...
-    private final UndertowSubsystemSchema schema;
-
-    AbstractUndertowSubsystemTestCase() {
-        this(UndertowSubsystemSchema.CURRENT);
-    }
+    protected final UndertowSubsystemSchema schema;
 
     AbstractUndertowSubsystemTestCase(UndertowSubsystemSchema schema) {
-        super(UndertowExtension.SUBSYSTEM_NAME, new UndertowExtension(), schema, UndertowSubsystemSchema.CURRENT);
+        super(UndertowExtension.SUBSYSTEM_NAME, new UndertowExtension(), schema, UndertowSubsystemSchema.CURRENT.get(schema.getStability()));
```
Comment (PrarthonaPaul) [inline undertow/src/test/java/org/wildfly/extension/undertow/AbstractUndertowSubsystemTestCase.java:40]: Do you have any tests that ensure that schema versions 13 and earlier are also parsed correctly? 
If you would like to to that without adding a new class, you can use `@RunWith(Parameterized.class)` tag at the beginning of the class and the parameters can be `EnumSet.allOf(UndertowSubsystemSchema.class)`

**Review by PrarthonaPaul — COMMENTED** [review]
Thanks @baranowb 
A few other comments: 
- Are you incrementing the SubsystemModel version? If not, it would be good to do that here: https://github.com/wildfly/wildfly/blob/2580d70ad57a5ff4f55ecd009c96eb552d2fdd75/undertow/src/main/java/org/wildfly/extension/undertow/UndertowSubsystemModel.java
- It would be good to add a transformer and tests to ensure that if you start the server at default stability using the preview schema, then the server fails to start. (Feel free to let me know if you have any questions). 
-  Does it make sense to add integration tests inside `wildfly/testsuite`?

File: undertow/src/test/java/org/wildfly/extension/undertow/AbstractUndertowSubsystemTestCase.java (hunk 2 above)
Comment (baranowb) [reply undertow/src/test/java/org/wildfly/extension/undertow/AbstractUndertowSubsystemTestCase.java:40]: AFAIR comment in code suggests its not done, neither RFE is about that :)

Comment (PrarthonaPaul) [reply undertow/src/test/java/org/wildfly/extension/undertow/AbstractUndertowSubsystemTestCase.java:40]: Sounds good!

File: undertow/src/main/resources/org/wildfly/extension/undertow/LocalDescriptions.properties [hunk 3]
```diff
@@ -374,6 +374,8 @@ undertow.handler.reverse-proxy.host.ssl-context=Reference to the SSLContext to b
 undertow.handler.reverse-proxy.host.security-realm.deprecated=Use the ssl-context attribute to reference a configured SSLContext directly.
 undertow.handler.reverse-proxy.host.enable-http2=If this is true then the proxy will attempt to use HTTP/2 to connect to the backend. If it is not supported it will fall back to HTTP/1.1.
 undertow.handler.reverse-proxy.max-retries=The number of times to attempt to retry a request if it fails. Note that if a request is not considered idempotent then it will only be retried if the proxy can be sure it was not sent to the backend server).
+undertow.handler.reverse-proxy.reuse-x-forwarded-header=Append(reuse existing) remote hop to x-forwarded-* headers, if not, previous value is discarded and remote hop becomes the value.
+undertow.handler.reverse-proxy.rewrite-host-header=Set host header to connection remote end and create x-forwarded-host header.
```
Comment (bstansberry) [inline undertow/src/main/resources/org/wildfly/extension/undertow/LocalDescriptions.properties:378]: "If true set the Host HTTP header to the proxy connection's remote end and create an x-forward-host header."

File: undertow/src/main/resources/org/wildfly/extension/undertow/LocalDescriptions.properties [hunk 4]
```diff
@@ -374,6 +374,8 @@ undertow.handler.reverse-proxy.host.ssl-context=Reference to the SSLContext to b
 undertow.handler.reverse-proxy.host.security-realm.deprecated=Use the ssl-context attribute to reference a configured SSLContext directly.
 undertow.handler.reverse-proxy.host.enable-http2=If this is true then the proxy will attempt to use HTTP/2 to connect to the backend. If it is not supported it will fall back to HTTP/1.1.
 undertow.handler.reverse-proxy.max-retries=The number of times to attempt to retry a request if it fails. Note that if a request is not considered idempotent then it will only be retried if the proxy can be sure it was not sent to the backend server).
+undertow.handler.reverse-proxy.reuse-x-forwarded-header=Append(reuse existing) remote hop to x-forwarded-* headers, if not, previous value is discarded and remote hop becomes the value.
```
Comment (bstansberry) [inline undertow/src/main/resources/org/wildfly/extension/undertow/LocalDescriptions.properties:377]: "If true append the remote hop value to the value of any existing x-forwarded-* headers. If false any previous header value is discarded and the remote hop becomes the value."

**Review by bstansberry — CHANGES_REQUESTED** [review]

**Review by fl4via — APPROVED** [review]

**Comment by bstansberry** [comment]
This was merged as part of #18002.

Thanks @baranowb @ropalka @PrarthonaPaul @fl4via

**Comment by bstansberry** [comment]
I re-opened this by mistake; closing again.

**Pull request closed without merging**